"My Spaces" search

andrewazores opened the pull request:
This PR introduces initial support for a search feature on the My Spaces page, as described in the relevant design issue [0]. A button is added to the My Spaces toolbar which summons a search dialog, containing a simple search term text field and submission button above an infinite scroll results list. Other features described in the design issue, such as search typeahead, the ability to request to join spaces, and the ability to sort the results list, are left implemented for future enhancements, since these will require additional enabling backend support features.

The invision designs for the feature are here: https://redhat.invisionapp.com/share/P7LNZNZUZ9Y#/screens

And this GIF is a quick demo of what the feature looks like as proposed in this PR:

![43343519-836f4d9e-91b4-11e8-9822-7e18ea0acfba](https://user-images.githubusercontent.com/3787464/44547446-239eca80-a6e9-11e8-835a-79b7f77032e7.gif)

Below are some more screenshots illustrating the feature.

Toolbar button added:
![screenshot_2018-08-23_15-03-31](https://user-images.githubusercontent.com/3787464/44546449-f00e7100-a6e5-11e8-9afb-b597b77dca44.png)

Initial state:
![screenshot_2018-08-23_15-03-58](https://user-images.githubusercontent.com/3787464/44546455-f7ce1580-a6e5-11e8-8756-d2630649ca14.png)

Empty state (no results for search):
![screenshot_2018-08-23_15-04-20](https://user-images.githubusercontent.com/3787464/44546458-fe5c8d00-a6e5-11e8-9260-9de84f44c9ae.png)

Results list:
![screenshot_2018-08-23_15-04-42](https://user-images.githubusercontent.com/3787464/44546465-0288aa80-a6e6-11e8-893d-59a34704df5f.png)


[0] https://github.com/openshiftio/openshift.io/issues/3655

alien-ike commented:
### Ike Plugins (test-keeper)

Thank you @andrewazores for this contribution!

<img align="left" src="https://raw.githubusercontent.com/arquillian/ike-prow-plugins/master/docs/images/arquillian_ui_failure_64px.png">

It appears that no tests have been added or updated in this PR.

Automated tests give us confidence in shipping reliable software. Please add some as part of this change.

If you are an admin or the reviewer of this PR and you are sure that no test is needed then you can use the command `/ok-without-tests` as a comment to make the status green.


For more information please head over to official [documentation](http://arquillian.org/ike-prow-plugins/#_test_keeper_plugin). You can find there how to configure the plugin.

rohitkrai03 commented:
@andrewazores We already use a `ng-select` component in add-collaborators-dialog [component](https://github.com/fabric8-ui/fabric8-ui/blob/master/src/app/space/settings/collaborators/add-collaborators-dialog/add-collaborators-dialog.component.html) which has all the APIs needed here as well. Do you think it would be a good idea to reuse that component for better consistency?

andrewazores commented:
@rohitkrai03 I'm not sure I understand. That component is used for searching and adding users as collaborators to a space. This WIP is for adding a dialog that adds the ability to search for spaces to the My Spaces page as a replacement for the View All Spaces page. How can the AddCollaboratorsDialog component be reused here?

rohitkrai03 commented:
@andrewazores I am not talking about reusing `AddCollaboratorsDialog` component, but reusing `ng-select` component as the search box. I see that you've created your own implementation of a search box for this PR.

andrewazores commented:
Ah, I see. Well, the search component here is just a plain `<input>` and a button next to it - I'm not sure that replacing this with `ng-select` makes sense at this point. Once there is support for typeahead then that would make sense IMO, since the functionality of the search bar would become similar to the one in the AddCollaborators dialog.

rohitkrai03 commented:
Okay, that makes sense.

andrewazores commented:
Removing WIP label as I've cleaned up the code and added tests. The state of the feature can be seen in this issue comment: https://github.com/openshiftio/openshift.io/issues/3655#issuecomment-408521519 . Other features such as sorting and typeahead will be future enhancements.

andrewazores commented:
@rohitkrai03 would you mind taking a look at this for me?

rohitkrai03 commented:
@andrewazores Sure, will take a look. Also, the parent issue has very big conversation thread so it would be better if you could give a proper description of what changes this PR is introducing. A few screenshots would also be helpful.

andrewazores commented:
@rohitkrai03 thanks, I've added some details and a GIF to the original PR description. I'll attach some screenshots as well.

andrewazores commented:
Just realized I left uncommitted work on my other computer, but it's just test additions. I'll get those merged and added back in ASAP.

andrewazores commented:
Tests updated

jiekang commented:
Do the API responses give you enough info to display something along the lines of number of results and number remaining to be shown?

jiekang commented:
Do you know why so many of the images are unset? I assumed it would default to that blue circle one but it looks otherwise.

![my-space](https://user-images.githubusercontent.com/5430520/44682506-940f5980-aa11-11e8-9a55-0e5ad6136a08.png)

andrewazores commented:
There is a number of results displayed. In my "Results list" screenshot, it says "134 items" in the top right. That's the total number of matching results. From there it's possible to calculate how many are expected in the results that have not yet been shown, but this was not included in the designs for the feature.

I don't know why those images don't load. The view just attempts to load the image specified by the query response. I can investigate if those are due to an empty image URL string or some similar case and either hide the image or load a default instead, though I would expect the query response to point to some reasonable default to begin with.

jiekang commented:
Ah my bad I missed the numbers. I feel like in most designs it's at the bottom right of the screen; not a big deal though.

jiekang commented:
Accidentally closed; sorry!

rohitkrai03 commented:
@andrewazores @jiekang Images being unset is a known issue and it is because Image URL is not set for a lot of users since by default backend doesn't set it while creating new user. We need to add default user image for cases where no `imageURL` is set.
Related - https://github.com/fabric8-ui/fabric8-ui/pull/3082

andrewazores commented:
Ah nice, thanks for the clarification. I'll add a commit on top that handles this case.

jiekang commented on src/app/profile/my-spaces/my-spaces-search-dialog/my-spaces-search-spaces-dialog-space-item/my-spaces-search-spaces-dialog-space-item.component.spec.ts:
```diff
@@ -0,0 +1,27 @@
+  });
+
+  it('should be instantiable', function(this: TestingContext): void {
+    expect(this.testedDirective).toBeDefined();
```
I know this is a pretty dumb component but could the test be expanded a bit? Maybe one to verify userAvatar/defaultAvatar is correctl based off imageUrl in space?

andrewazores replied:
I have a test like this prepared, but I'm now stuck investigating why it causes PhantomJS to crash...

jiekang commented on src/app/profile/my-spaces/my-spaces-search-dialog/my-spaces-search-spaces-dialog.component.html:
```diff
@@ -0,0 +1,83 @@
+        </button>
+      </div>
+      <div class="modal-body">
+
```
Some of the new lines look strange to me; I think I'd prefer not having them. Mainly the ones that are followed by an indented element. Ones followed by an element at the same indent level are okay.

jiekang commented on src/app/profile/my-spaces/my-spaces-search-dialog/my-spaces-search-spaces-dialog.component.spec.ts:
```diff
@@ -0,0 +1,319 @@
+        });
+    });
+  });
+
```
I think this newline can be removed.

jiekang commented on src/app/profile/my-spaces/my-spaces-search-dialog/my-spaces-search-spaces-dialog.component.ts:
```diff
@@ -0,0 +1,128 @@
+          }
+        });
+  }
+
```
Also strange newline (to me)

jiekang commented:
There's a minor nit that should be addressed later via another issue. The base page has a scrollbar. When the search box is brought up, the scrollbar disappears and when the box is exited from, it reappears. It causes a slight shift of the page leftwards/rightwards when that happens; it's pretty strange.

jiekang commented:
@andrewazores Hmm as soon as I hit merge I wonder if this is something that should be feature flagged... I don't really know the process for these things though. I wonder when the last time we moved a feature into production was.

andrewazores commented:
@jiekang good point. I'll ask someone with permissions to create a name for this feature toggle and I'll open a followup PR to toggle out the new search button.

andrewazores commented:
Regarding the scrollbar issue, I can reproduce this in Chromium but not Firefox. I don't think there's anything about the specific dialog I've introduced here that would cause this... so maybe it's just some issue with bsModal, its CSS/the overall page CSS, and/or browser-specific CSS quirks?

jiekang commented:
Yeah... I'll see if I can take a look at this soon too. Can a lowest priority issue be opened to track this?

andrewazores commented:
Opened: https://github.com/openshiftio/openshift.io/issues/4269
